File: backend/app/src/main/kotlin/net/animeta/backend/controller/v3/DeleteRecordController.kt
package net.animeta.backend.controller.v3

import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.security.CurrentUser
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class DeleteRecordController(
    private val recordRepository: RecordRepository
) {
    data class Params(
        val id: Int
    )
    data class Result(
        val ok: Boolean
    )

    @PostMapping("/v3/DeleteRecord")
    @Transactional
    fun handle(@RequestBody params: Params, @CurrentUser currentUser: User): Result {
        val record = recordRepository.findById(params.id).orElseThrow { ApiException.notFound() }
        if (currentUser.id != record.user.id) {
            throw ApiException.permissionDenied()
        }
        recordRepository.delete(record)
        return Result(true)
    }
}

File: backend-ts/src/controllers/delete_record.controller.ts
import { Body, Controller, HttpCode, Post } from "@nestjs/common";
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { RecordService } from "src/services/record.service";
import { Connection } from "typeorm";
import { ApiException } from "./exceptions";

type Params = {
  id: number;
}

type Result = {
  ok: boolean;
}

@Controller('/api/v4/DeleteRecord')
export class DeleteRecordController {
  constructor(
    private connection: Connection,
    private recordService: RecordService,
  ) {}

  @Post()
  @HttpCode(200)
  async handle(
    @Body() params: Params,
    @CurrentUser() currentUser: User,
  ): Promise<Result> {
    const record = await this.recordService.get(params.id)
    if (!record)
      throw ApiException.notFound()
    if (currentUser.id !== record.user_id)
      throw ApiException.permissionDenied()

    await this.connection.transaction(async em =>
      this.recordService.delete(em, record)
    )

    return { ok: true }
  }
}
